### PR TITLE
Pipeline test

### PR DESCRIPTION
**Pull request by LouisLeNezet**
Add nf-test to the pipeline level.
Should we move the github action to use it instead of `nextflow run main.nf -profile test` ?

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/phaseimpute/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/phaseimpute _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: workflows/phaseimpute/tests/test_all.nf.test [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+            params {
+                pipelines_testdata_base_path = 'https://raw.githubusercontent.com/nf-core/test-datasets/phaseimpute/data/'
+                outdir = "$outputDir"
+                max_cpus = 4
+                max_memory = '5.GB'
+            }
+        }
+
+        then {
+            assertAll(
```
Comment (atrigila) [inline workflows/phaseimpute/tests/test_all.nf.test:134]: Can we assert if the results before and after simulation are different? For example, see the number of lines of the original VCF vs. the downsized VCF.

File: workflows/phaseimpute/tests/test_all.nf.test.snap [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        "meta": {
+            "nf-test": "0.8.4",
+            "nextflow": "23.10.1"
+        },
+        "timestamp": "2024-05-17T19:34:26.554486719"
+    },
+    "Check test_sim": {
+        "content": [
+            [
+                "simulation/NA12878_D1_Rchr21_16570000-16610000.bam",
```
[the remark this answers is not in the document]
Comment (atrigila) [reply workflows/phaseimpute/tests/test_all.nf.test.snap:186]: I think that the output from simulation should be a single bam. If the original input was a single bam, why are we splitting per chromosome? 

I imagine a use case where someone would input `NA1278_30x.bam` and would like `--step simulate` to generate `NA12878_1x.bam`, so they could use it in `--step impute`.

File: workflows/phaseimpute/tests/test_all.nf.test.snap [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        "meta": {
+            "nf-test": "0.8.4",
+            "nextflow": "23.10.1"
+        },
+        "timestamp": "2024-05-17T21:48:07.027867847"
+    },
+    "Check test_validate": {
+        "content": [
+            [
+                "validation/NA12878_Dnull_P1000GP.s.norel_Cchr21_SNP.txt",
```
Comment (atrigila) [inline workflows/phaseimpute/tests/test_all.nf.test.snap:136]: Should validation be per chromosome? Do we want a whole vcf validation as well?

**Review by atrigila — COMMENTED** [review]
Thanks for adding nf-test to the pipeline! Do we have examples of other nf-core pipelines using nf-test for their pipeline test? And in GitHub actions?

File: workflows/phaseimpute/tests/test_all.nf.test.snap [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+                "validation/NA12878_Dnull_P1000GP.s.norel_Call_SNP.txt",
+                "validation/NA19401_Dnull_P1000GP.s.norel_Call_SNP.txt",
+                "validation/NA20359_Dnull_P1000GP.s.norel_Call_SNP.txt",
+                "validation/TestQuality.txt",
+                "validation/concat/NA12878_truth_concat.vcf.gz",
+                "validation/concat/NA12878_truth_concat.vcf.gz.tbi",
+                "validation/concat/NA19401_truth_concat.vcf.gz",
+                "validation/concat/NA19401_truth_concat.vcf.gz.tbi",
+                "validation/concat/NA20359_truth_concat.vcf.gz",
+                "validation/concat/NA20359_truth_concat.vcf.gz.tbi"
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply workflows/phaseimpute/tests/test_all.nf.test.snap:123]: I would have let them.
Those are the reference files used for comparison.

File: workflows/phaseimpute/tests/test_all.nf.test.snap [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+                "imputation/glimpse1/NA19401_Rchr21_16570000-16610000.phase.bcf.csi",
+                "imputation/glimpse1/NA19401_Rchr22_16570000-16610000.ligate.vcf.gz",
+                "imputation/glimpse1/NA19401_Rchr22_16570000-16610000.ligate.vcf.gz.csi",
+                "imputation/glimpse1/NA19401_Rchr22_16570000-16610000.phase.bcf.csi",
+                "imputation/glimpse1/NA20359_Rchr21_16570000-16610000.ligate.vcf.gz",
+                "imputation/glimpse1/NA20359_Rchr21_16570000-16610000.ligate.vcf.gz.csi",
+                "imputation/glimpse1/NA20359_Rchr21_16570000-16610000.phase.bcf.csi",
+                "imputation/glimpse1/NA20359_Rchr22_16570000-16610000.ligate.vcf.gz",
+                "imputation/glimpse1/NA20359_Rchr22_16570000-16610000.ligate.vcf.gz.csi",
+                "imputation/glimpse1/NA20359_Rchr22_16570000-16610000.phase.bcf.csi",
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply workflows/phaseimpute/tests/test_all.nf.test.snap:232]: Done

File: workflows/phaseimpute/tests/test_all.nf.test.snap [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+                "imputation/quilt/NA12878_Rchr22_16570000-16610000.impute.annotate.vcf.gz",
+                "imputation/quilt/NA12878_Rchr22_16570000-16610000.impute.annotate.vcf.gz.tbi",
+                "imputation/quilt/NA19401_Rchr21_16570000-16610000.impute.annotate.vcf.gz",
+                "imputation/quilt/NA19401_Rchr21_16570000-16610000.impute.annotate.vcf.gz.tbi",
+                "imputation/quilt/NA19401_Rchr22_16570000-16610000.impute.annotate.vcf.gz",
+                "imputation/quilt/NA19401_Rchr22_16570000-16610000.impute.annotate.vcf.gz.tbi",
+                "imputation/quilt/NA20359_Rchr21_16570000-16610000.impute.annotate.vcf.gz",
+                "imputation/quilt/NA20359_Rchr21_16570000-16610000.impute.annotate.vcf.gz.tbi",
+                "imputation/quilt/NA20359_Rchr22_16570000-16610000.impute.annotate.vcf.gz",
+                "imputation/quilt/NA20359_Rchr22_16570000-16610000.impute.annotate.vcf.gz.tbi",
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply workflows/phaseimpute/tests/test_all.nf.test.snap:165]: Done

File: workflows/phaseimpute/tests/test_all.nf.test.snap (hunk 2 above)
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply workflows/phaseimpute/tests/test_all.nf.test.snap:186]: What we need to decide is: do we downsample by chromosomes or by samples ?
Doing it by chromosomes allow parellization and by samples will take less process.
We could afterwards concatenate the chromosomes into one files but that will add a process.

File: workflows/phaseimpute/tests/test_all.nf.test.snap (hunk 3 above)
Comment (LouisLeNezet) [reply workflows/phaseimpute/tests/test_all.nf.test.snap:136]: This shouldn't be done by chromosomes.

Comment (LouisLeNezet) [reply workflows/phaseimpute/tests/test_all.nf.test.snap:136]: I'll make the appropriate modifications.

**Review by LouisLeNezet — COMMENTED** [review]
Need to put back concat_truth and delete samtools merge from main workflow

**Review by LouisLeNezet — COMMENTED** [review]
Ready for review

File: conf/test.config [hunk 7]
```diff
@@ -34,3 +34,7 @@ params {
     // Impute tools
     tools  = "glimpse1"
 }
+
+withName: 'NFCORE_PHASEIMPUTE:PHASEIMPUTE:VCF_CHUNK_GLIMPSE:GLIMPSE_CHUNK' {
+    ext.args   = ["--window-size 10000", "--window-count 400", "--buffer-size 5000", "--buffer-count 30"].join(' ')
+}
```
Comment (atrigila) [inline conf/test.config:40]: Why does it have a different config than the default config at `imputation_glimpse1.config`?

File: subworkflows/local/utils_nfcore_phaseimpute_pipeline/main.nf [hunk 8]
```diff
@@ -210,27 +210,22 @@ workflow PIPELINE_INITIALISATION {
...
-
     if (params.chunks) {
         ch_chunks = Channel
-        .fromSamplesheet("chunks")
-        .map {
-            meta, file ->
-                [ meta, file ]
-    }} else {
-        ch_chunks = [[[],[]]]
+            .fromSamplesheet("chunks")
```
Comment (atrigila) [inline subworkflows/local/utils_nfcore_phaseimpute_pipeline/main.nf:226]: I am not sure if the 
```
meta, file -> [ meta, file ]
```
should be removed?  I am going to open a new issue to add a test for this just to be sure.

File: subworkflows/local/vcf_concordance_glimpse2/main.nf [hunk 9]
```diff
@@ -6,8 +6,8 @@ include { GUNZIP                      } from '../../../modules/nf-core/gunzip'
 workflow VCF_CONCORDANCE_GLIMPSE2 {
 
     take:
-        ch_vcf_emul   // VCF file with imputed genotypes [ [id], vcf, csi]
-        ch_vcf_truth  // VCF file with truth genotypes   [ [id], vcf, csi]
+        ch_vcf_emul   // VCF file with imputed genotypes [ [id, panel, tool, chr], vcf, csi]
```
Comment (atrigila) [inline subworkflows/local/vcf_concordance_glimpse2/main.nf:9]: We have to take into account that it won't have a tool (unless we require them to do so) when users start from `--step validate`.

File: subworkflows/local/vcf_sites_extract_bcftools/main.nf [hunk 10]
```diff
@@ -27,6 +28,10 @@ workflow VCF_SITES_EXTRACT_BCFTOOLS {
     BCFTOOLS_QUERY(ch_panel_sites, [], [], [])
     ch_versions = ch_versions.mix(BCFTOOLS_QUERY.out.versions.first())
 
+    // Convert TSC to Stitch format ","" to "\t"
+    GAWK(BCFTOOLS_QUERY.out.output, [])
+    ch_versions = ch_versions.mix(GAWK.out.versions)
```
Comment (atrigila) [inline subworkflows/local/vcf_sites_extract_bcftools/main.nf:33]: This is ok but I am addressing it in a different way, as we are going to use as input a single `--posfile`, therefore, the changes need to be in stitch. I am working on this in my current PR.

File: subworkflows/local/vcf_phase_panel/main.nf
Comment (atrigila) [inline subworkflows/local/vcf_phase_panel/main.nf:1]: `VCF_PHASE_SHAPEIT5` does not work.

Comment (atrigila) [reply subworkflows/local/vcf_phase_panel/main.nf:1]: I will open an issue, please let's address this in a separate PR, not in this one.

**Review by atrigila — APPROVED** [review]
Great contribution! 🎉 Thanks for adding nf-test and improving glimpse1 so it can work with the new structure.

I did observe some minor issues but let's please address this in a different PR. 

I have my reservations regarding `BAM_REGION`, which I think should be an optional process (similar to the behavior of `--remove_samples`). Let's discuss the design of this module further so we can find a sweet spot between the default behaviour and a more customized option :)

File: conf/test.config (hunk 7 above)
Comment (LouisLeNezet) [reply conf/test.config:40]: This is done to force multiple chunks in the test configuration, to verify that the chunks are correctly propagate.

File: subworkflows/local/vcf_concordance_glimpse2/main.nf (hunk 9 above)
Comment (LouisLeNezet) [reply subworkflows/local/vcf_concordance_glimpse2/main.nf:9]: That's true. Normally the subworkflow does not depend of its presence in the meta map.
We should find a way to specify if it's optional or not in the comment.

File: subworkflows/local/vcf_sites_extract_bcftools/main.nf (hunk 10 above)
Comment (LouisLeNezet) [reply subworkflows/local/vcf_sites_extract_bcftools/main.nf:33]: Okay 👍

File: subworkflows/local/vcf_phase_panel/main.nf
Comment (LouisLeNezet) [reply subworkflows/local/vcf_phase_panel/main.nf:1]: I agree, this one is already too huge.

**Pull request merged**